File: bench/langs/queens.pl
%% N queens, ALL solutions counted -- backtracking search, which is the
%% one shape both languages express naturally: Prolog gets it from the
%% engine, Python writes the recursion by hand.
%%
%%   main(+N)   answer is the number of solutions (8 -> 92)

safe(_, _, []).
safe(Q, D, [P|Ps]) :-
    Q =\= P + D,
    Q =\= P - D,
    D1 is D + 1,
    safe(Q, D1, Ps).

place([], Qs, Qs).
place(Un, Placed, Qs) :-
    select(Q, Un, Un1),
    safe(Q, 1, Placed),
    place(Un1, [Q|Placed], Qs).

queens(N, Qs) :- numlist(1, N, Ns), place(Ns, [], Qs).

%% One rep is one whole search, and every rep answers the same count.
searches(0, _, C, C) :- !.
searches(K, N, _, C) :-
    findall(1, queens(N, _), L),
    length(L, C0),
    K1 is K - 1,
    searches(K1, N, C0, C).

main(N, Reps) :-
    searches(Reps, N, 0, C),
    format("answer(~w)~n", [C]).

File: bench/langs/queens.py
"""N queens, all solutions counted, Reps times -- the same backtracking."""
import sys

def safe(q, d, placed):
    for p in placed:
        if q == p + d or q == p - d:
            return False
        d += 1
    return True

def place(un, placed, out):
    if not un:
        out[0] += 1
        return
    for i, q in enumerate(un):
        if safe(q, 1, placed):
            place(un[:i] + un[i+1:], [q] + placed, out)

def main(n, reps):
    c = 0
    for _ in range(reps):
        out = [0]
        place(list(range(1, n + 1)), [], out)
        c = out[0]
    print("answer(%d)" % c)

if __name__ == "__main__":
    main(int(sys.argv[1]), int(sys.argv[2]))
